### ISSUE TITLE
Missing `eval-type-backport` as required dependency in 2.10.* Pydantic line for Python 3.8/3.9

### ISSUE THREAD

**Issue by potiuk** [pydantic/pydantic#10958]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Hello @Viicos (cc: @pierrejeambrun).

This is a follow-up after #10924 and another issue to add for #10910 I think.

While Airflow has some protection in places (constraints when installing airflow), there is an aftermath of the recent fix in handling base classes containing `from __future__ import annotations` in Python 3.8 and 3.9.  When the base class (like it is in multiple versions of Airflow) will use the new style, Pydantic will fail while importing the class. This is breaking existing installation (even installed with constraints) if a user will manually upgrade Pydantic to latest version (which is quite likely as upgrade of 3rd-party libraies might do it).

We already know from #10924 and discussed at lenght in https://github.com/apache/airflow/pull/44249#issuecomment-2493386351 - installing `eval-type-backport` solves the problem for old versions of Airflow. But Airlfow does not have it as dependency in any released version yet (it may have it in 2.10.4) - and obviously released Airflow versions cannot be modified, and it's not trivial for people to upgrade from 2.7 or 2.8 or 2.9 to 2.10.4 - even if we release new version of Airflow.

Another option for users is to manually install `eval-type-backport` - to mitigate the issue, but then, this is a manual installation that is not obvious and we have no way to communicate it to Airlfow users, and you are risking to be flooded with similar issues like #10924. 

However - there is an easy fix, that should not require you to change much - I think you should seriously consider adding `eval-type-backport` as a dependency of Pydantic 2.10.2+ for Python 3.8 and 3.9.. This is super easy  fix and it should not have any side effects. 

It seems that this is what you expect to be installed anyway in those cases and it seems that it will protect anyone who would have similar problems with any other similar cases. This is a very small library with no dependencies on it's own and it is actually doing what Pydantic expects - whenever someone uses base classes with new style unitons/list and `from __future__ import annotations` (which has been working before for Pydantic < 2.10). 

I think this way you might prevent many problems that will report to your repository - similarly as @sydney-runkle  who upgraded Pydantic to latest version to create similar issues.

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.1
        pydantic-core version: 2.27.1
          pydantic-core build: profile=release pgo=false
                 install path: /usr/local/lib/python3.9/site-packages/pydantic
               python version: 3.9.20 (main, Nov 12 2024, 23:14:42)  [GCC 12.2.0]
                     platform: Linux-6.10.11-linuxkit-aarch64-with-glibc2.36
             related packages: typing_extensions-4.12.2 fastapi-0.115.5 mypy-1.9.0
                       commit: unknown
```

**Comment by Viicos** [comment]
Thanks for opening this issue. I wasn't aware the added `eval-type-backport` dependency wasn't present in any released version of airflow (git blaming shown it was added a couple months ago so I naively assumed a release was made since then).

While adding the `eval-type-backport` as a required dependency for py<=3.10 shouldn't be an issue, I'd like to first clarify the original issue.

The airflow issue we encountered only happened when one of the bases of a model is defined in a separate module had a forward annotation using the new syntax, and thus failed to evaluate (more details in my [comment](https://github.com/apache/airflow/pull/44249#issuecomment-2491919741)). Note that the issue would have already occured if the base class were to be defined in the same module. For this reason, I believe (but might be wrong) that this shouldn't be an issue for many existing code bases.

---

On the other hand, because there's no airflow released version with the `eval-type-backport` dependency included and because the migration from older airflow versions might be tricky, we need to fix this.

Instead of including the `eval-type-backport` lib as a dependency, we'll most likely go with https://github.com/pydantic/pydantic/pull/10962, which will fix the issue by _not_ evaluating forward references for private fields, and will fix another similar report in the `pyairtable` library.

**Comment by potiuk** [comment]
> Note that the issue would have already occured if the base class were to be defined in the same module. For this reason, I believe (but might be wrong) that this shouldn't be an issue for many existing code bases.

It really depends. The big problem (and reason why we had it in Airlfow) was that previous versions of Pydantic were OK with that. So we had very little chance to even discover that we have a this issue (because it was not an issue).

I think it will happen for basilcally anyone who - so far - wanted to use PEP-563  consistently for their code, but they later added Pydantic, did not know about `backport-eval-type` and had to add exclusions for their classes to not force PEP 563. You mentioned before "lack of consistency" - but this is exactly the choice of PEP-563 and striving for consistency - and  applying ruff rule to consistently use it everywhere  lead to it **reall**. Then exception caused by Pydantic) was really triggering it, so pretty much everyone who strives for consistency would go for minimising the number of exclusions - and attempt to exclude only the "final" pydantic classes - even if they also had a base class similar to our -  they would not even know that there is base class that needs also an exclusion. Of course hard to say how many codebases will have similar problem, but this is a very plausible, and likely pattern a number of people could follow. Yes. There are a few "ifs" there but it's nothing "unexpected" / "wrong" we had done there.

> On the other hand, because there's no airflow released version with the eval-type-backport dependency included and because the migration from older airflow versions might be tricky, we need to fix this.

Thanks for responding to it :). Yeah - it would be simply dificult to handle it from our side, and while yes - this is one of the Hyrum's law / XKCD https://www.xkcd.com/1172/ cases, in reality, the fact that it's been OK for at least 1.5 years and many releases of Airlfow, makes it a candidate to at least workaround, knowing the disruptive effect it might have.

> Instead of including the eval-type-backport lib as a dependency, we'll most likely go with https://github.com/pydantic/pydantic/pull/10962, which will fix the issue by not evaluating forward references for private fields, and will fix another similar report in the pyairtable library (see https://github.com/pydantic/pydantic/pull/10962).

That will work for us. Thank you !

**Comment by potiuk** [comment]
Happy to test it once it's merged to main - we know know that it should work for the right gitHub syntax in our CI

**Comment by Viicos** [comment]
> I think it will happen for basilcally anyone who - so far - wanted to use PEP-563 consistently for their code and they have slightly more complex hierarchy of classes that want to mix-in some common behaviour in their pydantic objects.

You are right on this part, except that:
- this only broke private fields (i.e. prefixed by `_`), less common that "normal" fields. For "normal" fields, the exception would have been raised anyway in Pydantic < 2.10.
- the exception would still occur if the field annotation in one of the bases (e.g. the `logger` one in airflow) was only referencing builtins, e.g. `logger: int | str`. In this case, there's no need to provide any explicit namespace as `'int'` and `'str'` are automatically resolved, thanks to `eval` injecting `__builtins__`.

So the impact is indeed limited (but yes, hard to know how much it is, so far we got two reports — including the airflow one — of this issue).

Btw, thanks for sharing https://www.hyrumslaw.com/. I like these kind of articles related to API and library design. A lot of times, it is easier to respond to some user issue/request with a link to an article than repeating ourselves :)

**Comment by Viicos** [comment]
Ans yeah regarding PEP 563 and the runtime behavior of typing annotations in general, it is really challenging to correctly support everything [^1], and users (they are not to blame, as they are used to work with static type checkers, taking advantage of the AST analysis so that it is decoupled from the runtime behavior) are usually not aware of this.

If you are curious, you can take a look at [the implementation](https://github.com/alexmojaki/eval_type_backport/blob/main/eval_type_backport/eval_type_backport.py) of the `eval-type-backport`. It is interesting stuff, but a bit scary at the same time.

[^1]: Misc. examples: `List['Forward']` is converted to `List[ForwardRef('Forward')]` at runtime, while `list['Forward']` is left untouched. Typed dictionaries have their `__annotations__` converted to `ForwardRef` instances, which is actually painful to deal with for Pydantic. etc, etc.

**Comment by potiuk** [comment]
> If you are curious, you can take a look at [the implementation](https://github.com/alexmojaki/eval_type_backport/blob/main/eval_type_backport/eval_type_backport.py) of the eval-type-backport. It is interesting stuff, but a bit scary at the same time.

Yeah that eval is a bit of bandaid.

**Comment by Viicos** [comment]
> Happy to test #10962 once it's merged to main

It's not clear to me if the suggested syntax https://github.com/apache/airflow/pull/44260#discussion_r1852542551 worked in your CI, but if it does then in the meanwhile `pydantic @ git+https://github.com/pydantic/pydantic.git@private-fields-eval` should also work. But anyway, it should be merged later during the day.

**Comment by potiuk** [comment]
Runing the test here https://github.com/apache/airflow/pull/44339

**Comment by pierrejeambrun** [comment]
Glad to see things moving forward.

I admit that the issue is limited to a very specific use case, and I'm sorry we depend on this "implicit" interface but taht would be quite a pain for airflow to handle without this new release.

Thanks for your prompt response and handling of the matter @Viicos and most generally the pydantic team :)

**Comment by sydney-runkle** [comment]
Thanks @Viicos for your great work here, much appreciated!
